app/src/main/java/me/hufman/androidautoidrive/music/spotify/SpotifyAuthStateManager.kt

```diff
@@ -0,0 +1,159 @@
+package me.hufman.androidautoidrive.music.spotify
+
+import android.content.Context
+import android.content.SharedPreferences
+import android.util.Log
+import com.adamratzman.spotify.SpotifyException
+import net.openid.appauth.*
+import org.json.JSONException
+
+/**
+ * Spotify manager for an [AuthState] instance that is created from and mutates a JSON shared preferences
+ * file. If there are multiple instances of this class modifying the [AuthState], the current instance
+ * of the [AuthState] must manually be refreshed.
+ */
+class SpotifyAuthStateManager(context: Context) {
```

Comment:
Consider changing SpotifyAuthStateManager to receive a MutableAppSettings object instead of reimplementing SharedSettings on the given context. This will allow the unit tests to pass in a MockAppSettings object for easier unit tests.
In fact, with an AppSettingsObserver, you could detect a settings change and mark the AuthState as dirty and automatically `refreshCurrentState` on the next of any `get` that happens.

Reply:
Would the AppSettings be an appropriate place to store a serialized JSON file of the auth state? I like the idea of having something like that handle the IO of the shared prefs but I feel like it might be a little out of place there given it handles the settings and this data would kind of be tossed into there if that makes sense 🤔 . 

I do like the idea of having a dirty flag or something and then auto calling the refreshCurrentState() method on the next get(), not sure why I didn't initially do that 😅

Reply:
AppSettings is just a tiny wrapper on top of SharedPreferences, so tiny that it pretty much only deals with strings. I don't think there will be too much IO, and I don't mind too much about muddying the Settings category (there's a already few other AppSettings that aren't strictly user preferences too) :+1:

Reply:
I updated the SpotifyAuthStateManager to use AppSettings and also be a singleton due to some limitations/issues that I ran into using AppSettings for IO. Also this class makes more sense as a singleton as the issue of keeping several instances of the SpotifyAuthStateManager up to date and with the potential use of different context's for the AppSettings adds a lot of unneeded complexity to remain as a regular class